crates/iceberg/src/writer/partitioning/fanout_writer.rs

```diff
@@ -0,0 +1,386 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! This module provides the `FanoutWriter` implementation.
+
+use std::collections::HashMap;
+use std::marker::PhantomData;
+
+use async_trait::async_trait;
+
+use crate::spec::{PartitionKey, Struct};
+use crate::writer::partitioning::PartitioningWriter;
+use crate::writer::{DefaultInput, DefaultOutput, IcebergWriter, IcebergWriterBuilder};
+use crate::{Error, ErrorKind, Result};
+
+/// A writer that can write data to multiple partitions simultaneously.
+///
+/// Unlike `ClusteredWriter` which expects sorted input and maintains only one active writer,
+/// `FanoutWriter` can handle unsorted data by maintaining multiple active writers in a map.
+/// This allows writing to any partition at any time, but uses more memory as all writers
+/// remain active until the writer is closed.
+///
+/// # Type Parameters
+///
+/// * `B` - The inner writer builder type
+/// * `I` - Input type (defaults to `RecordBatch`)
+/// * `O` - Output collection type (defaults to `Vec<DataFile>`)
+#[derive(Clone)]
```

Comment:
Why we need this? Cloning a `FanoutWriter` is ambiguous, since it contains states like opened writers and data files.